misc-taurus: driver to communicate between cluster and cr7 applicaiton.

iusyk opened the pull request:
Sample of the usage

typedef struct {
    uint64_t   value;
    uint16_t   ioctl_cmd;
}taurus_cluster_data_t;

enum CtlIO_id{
    SPEED = 1,
    GEAR = 2,
    RPM = 3,
    TURN = 4,
    DOOR_OPEN = 5,
    FOG_LIGHTS_BACK = 6,
    FOG_LIGHTS_FRONT = 7,
    HIGH_BEAMS_LIGHT = 8,
    HIGH_ENGINE_TEMPERATURE = 9,
    LOW_BATTERY = 10,
    LOW_BEAMS_LIGHTS = 11,
    LOW_FUEL = 12,
    LOW_OIL = 13,
    LOW_TIRE_PRESSURE = 14,
    SEAT_BELT = 15,
    SIDE_LIGHTS = 16,
    BATTERY_ISSUE = 17,
    AUTO_LIGHTING_ON = 18,
    CLUSTER_ACTIVE = 255,
};

taurus_cluster_data_t data = {
	    .value = 100,
            .ioctl_cmd = (uint16_t)CtlIO_id::SPEED,/*set cluster-speed as 100*/
};
int f = open("/dev/cluster-taurus", O_WRONLY);

write(f, &data, sizeof(data));

close(f);

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>

iusyk committed:
misc-taurus: driver to communicate between cluster and cr7 applicaiton.

Sample of the usage
include <uapi/linux/r_taurus_cluster_protocol.h>

typedef struct {
    uint64_t   value;
    uint16_t   ioctl_cmd;
}taurus_cluster_data_t;

taurus_cluster_data_t data = {
	    .value = 100,
            .ioctl_cmd = CLUSTER_SPEED,/*set cluster-speed as 100*/
};
int f = open("/dev/cluster-taurus", O_WRONLY);

write(f, &data, sizeof(data));
/* second option */
ioctl(f, CLUSTER_SPEED, 100);

close(f);

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented:
Please run `checkpatch` (like `scripts/checkpatch.pl -g HEAD`), there are lots of errors.

lorc commented on drivers/misc/cldatadrv/r_taurus_bridge.h:
```diff
@@ -0,0 +1,140 @@
+#ifndef R_TAURUS_BRIDGE_H
+#define R_TAURUS_BRIDGE_H
+
+#ifdef __cplusplus
```
There is no C++ in the Linux kernel

lorc replied:
Why are you adding this and then removing in the next commit?

lorc commented on drivers/misc/cldatadrv/r_taurus_bridge.h:
```diff
@@ -0,0 +1,140 @@
+
+#ifndef R_TAURUS_BRIDGE_H
```
Looks like this is a copy of drivers/gpu/drm/rcar-rvgc/r_taurus_bridge.h

We don't need to copy definition. Move that file into some common place and include it for all TAURUS drivers

lorc replied:
Why are you adding this file and then removing it in the next patch?

lorc commented on drivers/misc/cldatadrv/r_taurus_cluster_protocol.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef R_TAURUS_CLUSTER_PROTOCOL_H
```
This file should go to the `include/uapi/linux` so userspace apps can include it as well. Check out other files in this directory as an example.

lorc commented on drivers/misc/cldatadrv/r_taurus_protocol_ids.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef _R_TAURUS_PROTOCOL_IDS_H_
```
We already have the same file in `drivers/gpu/drm/rcar-rvgc/`. No need to make copies. Just move the original file to some shared location

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+/*
```
Missing SPDX licence identifier

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+#include <linux/rpmsg.h>
+#include <linux/of_reserved_mem.h>
+#include <linux/atomic.h>
+#include <linux/skbuff.h>
```
Do you really need all those includes?

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+static const struct file_operations misc_fops = {
+	.owner = THIS_MODULE,
+	.open = misc_open,
+	.write = misc_write,
```
Looking at the API, `.write` is an invalid semantic for such operations. You need `ioctl` there.

lorc replied:
You need to remove `.write` completely. Write semantic is invalid in this case.

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+	struct taurus_event_list* event = NULL;
+	rcar_cluster_device_t *clusterdrv = NULL;
+
+	event = devm_kzalloc(&rpdev->dev, sizeof(*event), GFP_KERNEL);
```
I see no reason to use `devm_` allocator for short-lived buffers.

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+	}
+
+	msg.Id = cluster_taurus_get_uniq_id();
+	msg.Channel = 0x80;
```
Why `0x80`? Looks like magic number

iusyk replied:
No, this is the channel number has been defined in cr7 application to identify the source of the message

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+	
+	if (ret) {
+		dev_err(&rpdev->dev, "rpmsg_send failed: %d\n", ret);
+		goto end;
```
Don't you need to remove the event from the event list in this case?

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+	
+	if(wait_for_completion_interruptible(&event->completed) == -ERESTARTSYS) {
+		dev_err(&rpdev->dev, "send_msg: Interrupted while waiting taurus response (%d)\n", ret);
+		goto end;
```
Don't you need to remove the event from the event list in this case?

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+
+	taurus_cluster_data_t cluster_data = {
+		.value = 10,
+		.ioctl_cmd = 1,
```
What those values mean?

iusyk replied:
just ping to test communication, can be removed

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+	};
+	dev_info(&rpdev->dev, "cluster: send_msg: probe\n");
+
+	clusterdvc = kzalloc(sizeof(*clusterdvc), GFP_KERNEL);
```
....and this is the place where `devm_alloc` will be more suitable

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+	dev->minor = MISC_DYNAMIC_MINOR;
+
+	mutex_lock(&idr_lock);
+	ret = idr_alloc(&misc_idr, clusterdvc, 0, MAX_MISC_RPMSG, GFP_KERNEL);
```
I can't see where are you using `ret`. Why do you need IDR allocator at all?

iusyk replied:
ret usage 
if (ret < 0) {
		goto free_cluster_data;
	}

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+	rcar_cluster_device_t *data = dev_get_drvdata(&rpdev->dev);
+	
+	mutex_lock(&idr_lock);
+	idr_remove(&misc_idr, 0);
```
And there you use hardcoded IDR `0`.

I believe, you can just remove all this stuff, if you are not using it properly

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,330 @@
+		goto free_kbuf;
+	}
+	
+	data = (taurus_cluster_data_t*)kbuf;
```
If userspace try to wrote exactly 1 byte (or 2, or 3 bytes), you will have big problems.
This is one of the reasons why you need to use IOCTLs.

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef __RCAR_CLUSTER_DRV_H__
```
Definitions in this file are not used by other files, so you safely can remove this `.h` file and move all the definitions to the `.c` file

iusyk committed:
taurus: move common taurus .h files in common folder

The drivers misc/cldatadrv and gpu/drm/rcar-rvgc use the same files:
r_taurus_rvgc_protocol.h
r_taurus_bridge.h

It is reasonable move them into include/linux to share between the drivers.

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented on include/uapi/linux/r_taurus_cluster_protocol.h:
```diff
@@ -0,0 +1,40 @@
+# define __packed       __attribute__((__packed__))
+#endif
+
+#include <linux/r_taurus_bridge.h>
```
I am not sure that this include will be available to userpsace. Could you please check?

iusyk replied:
this is because it took the file from rvgc :( and did not check

lorc commented on drivers/gpu/drm/rcar-rvgc/rcar_rvgc_pipe.c:
```diff
@@ -21,7 +21,7 @@
 #include "rcar_rvgc_taurus.h"
 #include "r_taurus_rvgc_protocol.h"
-#include "r_taurus_bridge.h"
+#include "linux/r_taurus_bridge.h"
```
`#include <linux/r_taurus_bridge.h>`

lorc commented on drivers/gpu/drm/rcar-rvgc/r_taurus_rvgc_protocol.h:
```diff
@@ -2,8 +2,8 @@
-#include "r_taurus_bridge.h"
-#include "r_taurus_protocol_ids.h"
+#include "linux/r_taurus_bridge.h"
+#include "linux/r_taurus_protocol_ids.h"
```
`#include <linux/r_taurus_bridge.h>`
`#include <linux/r_taurus_protocol_ids.h>`

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,389 @@
+#include <linux/wait.h>
+#include <linux/miscdevice.h>
+#include <linux/rpmsg.h>
+#include "uapi/linux/r_taurus_cluster_protocol.h"
```
`#include <uapi/linux/r_taurus_cluster_protocol.h>`

iusyk replied:
regarding cpp , this is because it took the file from rvgc :( and did not check

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,389 @@
+
+static long misc_ioctl(struct file *p_file, unsigned int cmd, unsigned long arg)
+{
+	unsigned int available_commands[] = {
```
Make it `static const`. No need to allocate it on the stack for each call.

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,389 @@
+		CLUSTER_ACTIVE
+	};
+	unsigned int index = 0;
+	unsigned int len = sizeof(available_commands)/sizeof(unsigned int);
```
There is `ARRAY_SIZE` in `<linux/kernel.h>`

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,389 @@
+	if (!kbuf)
+		return -ENOMEM;
+
+	data = (taurus_cluster_data_t *)kbuf;
```
Why do you need `kbuf`? Why not 

`data = kzalloc(sizeof(taurus_cluster_data_t), GFP_KERNEL);`?

lorc replied:
Also, this is short-lived small buffer. You can allocate in right on stack.

lorc replied:
Moreover, you can remove `sizeof(taurus_cluster_data_t)` at all. Just pass `cmd` and `arg` as a two separate parameters to `send_msg`.  You will unpack this structure there anyways. No need for extra memory operations.

lorc commented on drivers/misc/cldatadrv/rcar_cluster_drv.c:
```diff
@@ -0,0 +1,389 @@
+#define RCAR_CLUSTER_NAME     "rcar-cluster-drv"
+
+#define RCAR_IO_SPEED  1
+#define RCAR_IO_GEAR   2
```
I can't see where this two defines are used

lorc commented:
1. Please fix the checkpatch issues:

```
-----------------------------------------------------------------------------------------------
Commit fa7315e50636 ("misc-taurus: driver to communicate between cluster and cr7 applicaiton.")
-----------------------------------------------------------------------------------------------
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#62: 
new file mode 100644

WARNING: Missing or malformed SPDX-License-Identifier tag in line 1
#90: FILE: drivers/misc/cldatadrv/r_taurus_protocol_ids.h:1:
+#ifndef _R_TAURUS_PROTOCOL_IDS_H_

ERROR: code indent should use tabs where possible
#155: FILE: drivers/misc/cldatadrv/rcar_cluster_drv.c:50:
+        R_TAURUS_ResultMsg_t hdr;$

WARNING: please, no spaces at the start of a line
#155: FILE: drivers/misc/cldatadrv/rcar_cluster_drv.c:50:
+        R_TAURUS_ResultMsg_t hdr;$
```

2. Move "taurus: move common taurus .h files in common folder" to the first patch in the series. Because right now "misc-taurus: driver to communicate between cluster and cr7 applicaiton." adds  r_taurus_protocol_ids.h and rcar_cluster_drv.c    but then "taurus: move common taurus .h files in common folder" removes them. This is incorrect. Better to move the files first, and then use them on new locations. 

3. Squash "cldatadrv: taurus cluster protocol declarations to uapi" into "misc-taurus: driver to communicate between cluster and cr7 applicaiton.".

4. Stop marked issues "resolved" without resolving them. I asked to remove `write` method because it is semantically incorrect in this driver.

lorc reviewed (approved):
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`
